test/javascripts/lib/utilities-test.js.es6

```diff
@@ -186,3 +187,25 @@ QUnit.test("fillMissingDates", assert => {
     "it returns a JSON array with 31 dates"
   );
 });
+
+QUnit.test("inCodeBlock", assert => {
+  const text =
+    "000\n\n```\n111\n```\n\n000\n\n`111 111`\n\n000\n\n[code]\n111\n[/code]\n\n    111\n\t111\n\n000`111";
+
+  for (let i = 0; i < text.length; ++i) {
+    assert.equal(inCodeBlock(text, i), text[i] === "1");
+  }
+});
+
+QUnit.test("inCodeBlock - runs fast", assert => {
+  const phrase = "Lorem ipsum dolor sit amet, consectetur adipiscing elit.";
+  const text = `${phrase}\n\n\`\`\`\n${phrase}\n\`\`\`\n\n${phrase}\n\n\`${phrase}\n${phrase}\n\n${phrase}\n\n[code]\n${phrase}\n[/code]\n\n${phrase}\n\n    ${phrase}\n\n\`${phrase}\`\n\n${phrase}`;
+
+  const start = performance.now();
+  inCodeBlock(text, text.length);
+  const end = performance.now();
+
+  // This runs in 'keyUp' event handler so it should run as fast as
+  // possible. It should take less than 1ms for the test text.
+  assert.ok(end - start < 10);
```

Comment:
From my perspective this PR is good to merge, but I am a bit concerned about an assertion for performance, a single GC can through it off so this is ultra likely to trip. I think we should skip this test for now despite the awesome intent.

Maybe we should start thinking about a "performance" front end test suite longer term that handles this and more.

At the integration level this becomes mega interesting, maybe "typing in the composer does not lead to lag" type of integration smoke test. Probably worth discussing with @ZogStriP  and @eviltrout

Reply:
I tried to take into account the system noise by ensuring that the limit is 10 times of what I am seeing on my laptop. Something else I could do to consolidate this test is to run the code 10 times and take the minimum execution time. I doubt the GC can impact more than a few of these runs.

Measuring the minimum time makes sense because ultimately the execution time is a sum of the "real" execution time and noise. By taking the minimum, we are also taking the minimum noise, because in theory the execution time should stay constant. In other words, `time = real + noise`, `min(time) = min(real + noise)`, but `min(real) ~= real` so `min(time) = real + min(noise)`.

Anyway, for now, I will remove this test. I agree that it would be helpful to implement a framework for this kind of test and also to write tests for vital components such as the composer or topic view. Chrome kind of does this by `console.warn`ing the user if an event handler takes too much to run.

Reply:
I definitely think this should be a different class of test like smoke, because I guarantee we will see it fail in places we never expected it to otherwise.

As a future project we could add a "speed regression" suite that lets us know if things get slower.